File: 75-sort-colors/75.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
from typing import List


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return
        r, b, cur = 0, len(nums) - 1, 0
        while cur <= b and r < b:
            if nums[cur] == 0:
                nums[cur], nums[r] = nums[r], nums[cur]
                cur += 1
                r += 1
            elif nums[cur] == 2:
                nums[cur], nums[b] = nums[b], nums[cur]
                b -= 1
            else:
                cur += 1


if __name__ == '__main__':
    sol = Solution()
    nums = [2, 0, 2, 1, 1, 0]
    sol.sortColors(nums)
    print(nums)
    nums = [2, 0, 1]
    sol.sortColors(nums)
    print(nums)
    nums = [0]
    sol.sortColors(nums)
    print(nums)
    nums = [1]
    sol.sortColors(nums)
    print(nums)

File: shoope/75-sort-colors/75.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;

class Solution
{
public:
    void sortColors(vector<int> &nums)
    {
        int len = nums.size();
        if (len == 0)
        {
            return;
        }

        int l = 0, r = len - 1, cur = 0;

        while (cur <= r && l < r)
        {
            if (nums[cur] == 0)
            {
                swap(nums[cur], nums[l]);
                ++l;
                ++cur;
            }
            else if (nums[cur] == 2)
            {
                swap(nums[cur], nums[r]);
                --r;
            }
            else
            {
                ++cur;
            }
        }
    }
};

int main(int argc, char **args)
{
    Solution sol = Solution();
    vector<int> nums = {2, 0, 2, 1, 1, 0};
    sol.sortColors(nums);
    for (auto a : nums)
    {
        cout << a << " ";
    }
    cout << endl;
    nums = {2, 0, 1};
    sol.sortColors(nums);
    for (auto a : nums)
    {
        cout << a << " ";
    }
    cout << endl;
    nums = {0};
    sol.sortColors(nums);
    for (auto a : nums)
    {
        cout << a << " ";
    }
    cout << endl;
    nums = {1};
    sol.sortColors(nums);
    for (auto a : nums)
    {
        cout << a << " ";
    }
    cout << endl;
    return 0;
}